spring-rabbit/src/main/java/org/springframework/amqp/rabbit/core/AsyncRabbitTemplate.java

```diff
@@ -0,0 +1,583 @@
+ * published.
+ * <p>
+ * Returned (undeliverable) request messages are presented as a
+ * {@link AmqpMessageReturnedException} cause of an {@link ExecutionException}.
+ * <p>
+ * Internally, the template uses a {@link RabbitTemplate} and a
+ * {@link SimpleMessageListenerContainer} either provided or constructed internally.
+ * If an external {@link RabbitTemplate} is provided and confirms/returns are enabled,
+ * it must not previously have had callbacks registered because this object needs to
+ * be the callback.
+ *
+ * @author Gary Russell
+ * @since 1.6
+ *
+ */
+public class AsyncRabbitTemplate<C> implements SmartLifecycle, MessageListener, ReturnCallback, ConfirmCallback {
+
+	private final Log logger = LogFactory.getLog(this.getClass());
+
+	private final RabbitTemplate template;
+
+	private final SimpleMessageListenerContainer container;
+
+	private final String replyAddress;
+
+	@SuppressWarnings("rawtypes")
+	private final ConcurrentMap<String, RabbitFuture> pending = new ConcurrentHashMap<String, RabbitFuture>();
+
+	private volatile boolean running;
+
+	private volatile boolean enableConfirms;
+
+	private int phase;
+
+	private boolean autoStartup = true;
+
+	private Charset charset = Charset.forName("UTF-8");
+
+	/**
+	 * Construct an instance using the provided arguments. Replies will be
+	 * routed to the default exchange using the reply queue name as the routing
+	 * key.
+	 *
+	 * @param connectionFactory the connection factory.
+	 * @param exchange the default exchange to which requests will be sent.
+	 * @param routingKey the default routing key.
+	 * @param replyQueue the name of the reply queue to listen for replies.
+	 */
+	public AsyncRabbitTemplate(ConnectionFactory connectionFactory, String exchange, String routingKey,
+			String replyQueue) {
+		this(connectionFactory, exchange, routingKey, replyQueue, null);
+	}
+
+	/**
+	 * Construct an instance using the provided arguments. If 'replyAddress' is null,
+	 * replies will be routed to the default exchange using the reply queue name as the
+	 * routing key. Otherwise it should have the form exchange/routingKey and must
+	 * cause messages to be routed to the reply queue.
+	 *
+	 * @param connectionFactory the connection factory.
+	 * @param exchange the default exchange to which requests will be sent.
+	 * @param routingKey the default routing key.
+	 * @param replyQueue the name of the reply queue to listen for replies.
+	 * @param replyAddress the reply address (exchange/routingKey).
+	 */
+	public AsyncRabbitTemplate(ConnectionFactory connectionFactory, String exchange, String routingKey,
+			String replyQueue, String replyAddress) {
+		Assert.notNull(connectionFactory, "'connectionFactory' cannot be null");
+		this.template = new RabbitTemplate(connectionFactory);
+		Assert.notNull(routingKey, "'exchange' cannot be null, use \"\" for the default exchange");
+		this.template.setExchange(exchange == null ? "" : exchange);
```

Comment:
The previous `notNull` says correct words, but it is is for the `routingKey`. Easy to fix though.
But this ternary operator does confuse me.
What is the reason in that `notNull`, if we just treat `null` as an empty String ?

I prefer to place all asserts in the beginning to avoid object creation, but that is just IMO.